tools/clang/lib/SPIRV/GlPerVertex.cpp

```diff
@@ -0,0 +1,751 @@
+  }
+  return decl->getType();
+}
+} // anonymous namespace
+
+GlPerVertex::GlPerVertex(const hlsl::ShaderModel &sm, ASTContext &context,
+                         ModuleBuilder &builder, TypeTranslator &translator)
+    : shaderModel(sm), astContext(context), theBuilder(builder),
+      typeTranslator(translator), inIsGrouped(true), outIsGrouped(true),
+      inBlockVar(0), outBlockVar(0), inClipVar(0), inCullVar(0), outClipVar(0),
+      outCullVar(0), inArraySize(0), outArraySize(0), inClipArraySize(1),
+      outClipArraySize(1), inCullArraySize(1), outCullArraySize(1) {}
+
+void GlPerVertex::generateVars(uint32_t inArrayLen, uint32_t outArrayLen) {
+  // Calling this method twice is an internal error.
+  assert(inBlockVar == 0);
+  assert(outBlockVar == 0);
+
+  inArraySize = inArrayLen;
+  outArraySize = outArrayLen;
+
+  switch (shaderModel.GetKind()) {
+  case hlsl::ShaderModel::Kind::Vertex:
+    outBlockVar = createBlockVar(/*asInput=*/false, 0);
+    break;
+  case hlsl::ShaderModel::Kind::Hull:
+    inBlockVar = createBlockVar(/*asInput=*/true, inArraySize);
+    outBlockVar = createBlockVar(/*asInput=*/false, outArraySize);
+    break;
+  case hlsl::ShaderModel::Kind::Domain:
+    inBlockVar = createBlockVar(/*asInput=*/true, inArraySize);
+    outBlockVar = createBlockVar(/*asInput=*/false, 0);
+    break;
+  case hlsl::ShaderModel::Kind::Geometry:
+    inBlockVar = createBlockVar(/*asInput=*/true, inArraySize);
+    if (!outClipType.empty())
+      outClipVar = createClipDistanceVar(/*asInput=*/false, outClipArraySize);
+    if (!outCullType.empty())
+      outCullVar = createCullDistanceVar(/*asInput=*/false, outCullArraySize);
+    outIsGrouped = false;
+    break;
+  case hlsl::ShaderModel::Kind::Pixel:
+    if (!inClipType.empty())
+      inClipVar = createClipDistanceVar(/*asInput=*/true, inClipArraySize);
+    if (!inCullType.empty())
+      inCullVar = createCullDistanceVar(/*asInput=*/true, inCullArraySize);
+    inIsGrouped = false;
+    break;
+  }
+}
+
+llvm::SmallVector<uint32_t, 4> GlPerVertex::getStageInVars() const {
+  llvm::SmallVector<uint32_t, 4> vars;
+  if (inIsGrouped) {
+    if (inBlockVar)
+      vars.push_back(inBlockVar);
+  } else {
+    if (inClipVar)
+      vars.push_back(inClipVar);
+    if (inCullVar)
+      vars.push_back(inCullVar);
+  }
+
+  return vars;
+}
+
+llvm::SmallVector<uint32_t, 4> GlPerVertex::getStageOutVars() const {
+  llvm::SmallVector<uint32_t, 4> vars;
+  if (outIsGrouped) {
+    if (outBlockVar)
+      vars.push_back(outBlockVar);
+  } else {
+    if (outClipVar)
+      vars.push_back(outClipVar);
+    if (outCullVar)
+      vars.push_back(outCullVar);
+  }
+
+  return vars;
+}
+
+bool GlPerVertex::recordClipCullDistanceDecl(const DeclaratorDecl *decl,
+                                             bool asInput) {
+  const QualType type = getTypeOrFnRetType(decl);
+
+  if (type->isVoidType())
+    return true;
+
+  return doClipCullDistanceDecl(decl, type, asInput);
```

Comment:
Is there a reason to have `doClipCullDistanceDecl` separately? Couldn't you add the 'isVoid` condition to the function below and basically have just 1 recursive function?

Reply:
i see. No need to change.